llvm/lib/Transforms/Utils/LoopIterCountAssumptions.cpp

```diff
@@ -0,0 +1,251 @@
+// picked up by Loop Rotate to remove Loop Guards.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/Transforms/Utils/LoopIterCountAssumptions.h"
+#include "llvm/Analysis/AliasAnalysis.h"
+#include "llvm/Analysis/AssumptionCache.h"
+#include "llvm/Analysis/MemorySSA.h"
+#include "llvm/Analysis/MemorySSAUpdater.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+#include "llvm/Passes/PassBuilder.h"
+#include "llvm/Support/Casting.h"
+#include "llvm/Transforms/Scalar/LICM.h"
+#include "llvm/Transforms/Utils/LoopUtils.h"
+#include "llvm/Transforms/Utils/ScalarEvolutionExpander.h"
+
+#define DEBUG_TYPE "loop-iter-count-assumptions"
+
+using namespace llvm;
+
+namespace {
+
+std::string getFunctionAndBlockNames(const BasicBlock &BB) {
+  return BB.getParent()->getName().str() + " " + BB.getName().str();
+}
+
+/// Return the Branch Compare Instruction of CurrentLoop if the Loop is well
+/// formed and this pass can process the Predicate
+ICmpInst *getLoopCmpInst(const Loop &CurrentLoop) {
+
+  if (CurrentLoop.isRotatedForm()) {
+    LLVM_DEBUG(dbgs() << "Loop already in rotated form. Will not add Loop "
+                         "Iteration Count assumptions.\n");
+    return nullptr;
+  }
+
+  /// Check that the loop has a single Exiting Block. If the CurrentLoop
+  /// has multiple Exiting Blocks, ExitBB will be a nullptr
+  auto *ExitBB = CurrentLoop.getExitingBlock();
+  if (!ExitBB)
+    return nullptr;
+
+  BranchInst *BI = dyn_cast<BranchInst>(ExitBB->getTerminator());
+  if (!BI)
+    return nullptr;
+
+  ICmpInst *LoopCmpInstr = dyn_cast<ICmpInst>(BI->getCondition());
+  if (!LoopCmpInstr)
+    return nullptr;
+
+  LLVM_DEBUG(dbgs() << "Condition Found: " << *LoopCmpInstr << "\n");
+  return LoopCmpInstr;
+}
+
+bool hasVariableStepSize(Value &Op, ScalarEvolution &SE) {
+  const SCEVAddRecExpr *AddRec =
+      dyn_cast_or_null<SCEVAddRecExpr>(SE.getSCEV(&Op));
+  if (!AddRec)
+    return false;
+
+  const SCEV *StepSize = AddRec->getStepRecurrence(SE);
+  return !isa<SCEVConstant>(StepSize);
+}
+
+/// Return the AddRecExpr evaluated at Iteration \p IterCount if an
+/// AddRecExpr can be extracted, otherwise return loop invariant Value of \p Op
+Value *expandValueAtIteration(Value *Op, ScalarEvolution &SE,
+                              SCEVExpander &Expander,
+                              Instruction *InsertionPoint, Loop *CurrentLoop,
+                              int64_t IterCount) {
+  const SCEVAddRecExpr *AddRec =
+      dyn_cast_or_null<SCEVAddRecExpr>(SE.getSCEV(Op));
+  if (AddRec) {
+    const SCEV *IterSCEV =
+        AddRec->evaluateAtIteration(SE.getConstant(APInt(32, IterCount)), SE);
+
+    // Copy Overflow Flags to SCEV
+    SCEV::NoWrapFlags NWF = AddRec->getNoWrapFlags(
+        SCEV::NoWrapFlags(/*Mask=*/SCEV::FlagNUW | SCEV::FlagNSW));
+
+    // IterSCEV can either be an AddExpr or simplify to a MulExpr
```

Comment:
`or simplify to a MulExpr`, is this really true? Do you have a test that shows this example?

Reply:
yes, for example this unit test `LoopVariantUpperBound`

Reply:
I wasn't sure if adding a reference to a unittest would be appropriate

Reply:
Oh right, that makes sense. Somehow I read it as "recursive" `MulExpr`, but this here simply represents a multiplication of two SCEVs.